Support partitioned cookies in Clear-Site-Data: Cookies

sysrqb opened the pull request:
#### 78e86d4b6b77da06ba4eac1a2e4c6ac59300dcda
<pre>
Support partitioned cookies in Clear-Site-Data: Cookies
<a href="https://bugs.webkit.org/show_bug.cgi?id=283552">https://bugs.webkit.org/show_bug.cgi?id=283552</a>
<a href="https://rdar.apple.com/140403030">rdar://140403030</a>

Reviewed by Sihui Liu.

This is another step toward supporting opt-in partitioned cookies. When a
top-level site requests deleting cookies using the Clear-Site-Data header, we
should delete only the site&apos;s cookies. When a cross-site subframe requests
deleting cookies using the header, then we should delete only the cookies in
the partition.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/clear-site-data/set-and-clear-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/clear-site-data/set-and-clear-cookies.https.html: Added.
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::cookiePartitionIdentifier):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::deleteCookies):

Canonical link: <a href="https://commits.webkit.org/287547@main">https://commits.webkit.org/287547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb915d16ab0c5eb74ad0528657dc413c3515ce64

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58810 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33208 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84334 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30811 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81918 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7100 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62381 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20219 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82877 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52438 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72680 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42688 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26829 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29259 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70905 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27300 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85760 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7040 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4933 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-pause-networkState.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70640 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68520 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69881 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13885 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12804 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12381 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6999 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12577 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6859 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10370 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -616,9 +616,19 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
+        if (isMainFrame) {
+            if (String(cookie.domain) == domain && !cookie._storagePartition)
+                return true;
+            if (!cachePartition.isNull() && equalIgnoringNullity(String(cookie._storagePartition), cachePartition))
```
Does this mean a first-party response can clear data for cookies set by a third-party response?
Is this specified in spec?  I don't think this change is consistent with other storage types.
For the other types, the delete operation with `ClientOrigin` (e.g. {OriginA, OriginA}) will only clear data in that partition (e.g. OriginA/OriginA, not OriginA/OriginB).

sysrqb replied:
Unless I'm misinterpreting the proposed spec change, yes the first party can clear third-party partitioned cookies: https://datatracker.ietf.org/doc/html/draft-cutler-httpbis-partitioned-cookies#name-partitioned-cookies-and-cle

szewai replied:
Per discussion offline, we should limit the deletion to its own partition.

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -616,9 +616,18 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
+            if (String(cookie.domain) == domain && !cachePartition.isNull() && equalIgnoringNullity(String(cookie._storagePartition), cachePartition))
+                return true;
+        }
+        return false;
```
Can we simplify it as:
```
auto cachePartition = origin.topOrigin == origin.clientOrigin ? String { } : cookiePartitionIdentifier(origin.topOrigin);
auto domain = origin.clientOrigin.domain;
deleteCookiesMatching([&domain, & cachePartition](auto *cookie) { 
    return equalIgnoringNullity(cachePartition, String(cookie._storagePartition)) && domain == String(cookie.domain);
});

```

sysrqb replied:
It is possible that the top frame can set a partitioned cookie, and we should clear that, as well. So for the `origin.topOrigin == origin.clientOrigin` case, we'll compare with both the null string and the cookiePartitionIdentifier.

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -210,6 +210,17 @@ String NetworkStorageSession::cookiePartitionIdentifier(const ResourceRequest& r
+        return cookiePartitionIdentifier(origin.toURL());
+    }, [](const ProcessQualified<OpaqueOriginIdentifier>&) {
+        return String { };
+    });
```
Opaque origin gets empty url so you could just return `cookiePartitionIdentifier(origin.toURL())` (and this function might be unnecessary if you could do `origin.toURL()` in the caller).

sysrqb replied:
I'm not sure we should ever have an opaque origin in this case, but that is a good point.

szewai reviewed (approved):
LGTM with nits

Commit:
Support partitioned cookies in Clear-Site-Data: Cookies
https://bugs.webkit.org/show_bug.cgi?id=283552
rdar://140403030

Reviewed by Sihui Liu.

This is another step toward supporting opt-in partitioned cookies. When a
top-level site requests deleting cookies using the Clear-Site-Data header, we
should delete only the site's cookies. When a cross-site subframe requests
deleting cookies using the header, then we should delete only the cookies in
the partition.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/clear-site-data/set-and-clear-cookies.https-expected.txt: Added.
* LayoutTests/http/tests/clear-site-data/set-and-clear-cookies.https.html: Added.
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::cookiePartitionIdentifier):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::deleteCookies):

Canonical link: https://commits.webkit.org/287547@main
